Publicize: Add REST API

ockham opened the pull request:
Replaces #10372. Based on #10396. ~Rebase on `master` after merging #10396!~ (done)

#### Changes proposed in this Pull Request:

Add a REST API for Publicize.

#### Testing instructions:

See companion PR https://github.com/Automattic/wp-calypso/pull/28102.

This doesn't yet fully replicate the non-Gutenberg UI (e.g. no warning message next to a Facebook connection; publicize message placeholder; expansion behavior). I'd like to tackle those in a subsequent iteration.

#### Proposed changelog entry for your changes:

* Add a REST API for Publicize.

ockham committed:
Copy class-jetpack-publicize-gutenberg.php from #10372 to rest-api.php

ockham committed:
Rename Jetpack_Publicize_Gutenberg to Publicize_REST_API

ockham committed:
Remove obsolete JS globals

ockham commented on modules/publicize/rest-api.php:
```diff
@@ -0,0 +1,330 @@
+		 *
+		 * @since 5.9.1
+		 */
+		register_rest_route( 'publicize/', '/connections', array(
```
Yep, right now it's only copy-pasta from #10372's `class-jetpack-publicize-gutenberg.php`. I'm planning on moving the endpoint registrations into `class.core-rest-api-endpoints.php` since that seems to be where all the other endpoints are.

However, turns out that they're all in the `jetpack/v4` namespace, so I was inclined to stick with that. I'm not sure what the overall plans are for these...

ockham committed:
Add Publicize REST API Fields to Post endpoint

ockham committed:
Move Publicize Route Definitions to class.core-rest-api-endpoints.php

ockham committed:
Move process_publicize_from_rest() to Publicize class

ockham committed:
Add tests for process_publicize_from_rest()

ockham committed:
Remove genericons

ockham committed:
Add get_publicize_conns_test_results() function

ockham committed:
Add JSDoc to get_publicize_conns_test_results

tyxla commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -793,6 +816,17 @@ public static function activate_plugins_permission_check() {
 	}
 
+	/**
+	 * Verify that user can publish posts.
```
Nit: here and in the `return` we say just "user", but it'd probably make sense to make it "the current user".

tyxla commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -382,6 +384,27 @@ public static function register_endpoints() {
 		) );
 
+		// Publicize: get connection list data for current user.
+		register_rest_route( 'jetpack/v4', '/publicize/connections', array(
```
I'm not sure how we can have the `/jetpack/v4` endpoints accessible in Jetpack, Calypso and WP.com at the same time (also mentioned here https://github.com/Automattic/wp-calypso/pull/28102#pullrequestreview-168743801). Perhaps @mdawaffe will have some suggestions.

ockham replied:
Yeah, fair point. I was striving for consistency with existing endpoints in Jetpack's `class.core-rest-api-endpoints.php` (which are all in `/jetpack/v4`). I think this is part of a bigger question how all of those endpoints are going to work.

mdawaffe replied:
I don't think we have a firm decision here.

My opinion:
* `jetpack/v4` should be for things dealing with the settings for the Jetpack plugin itself: activate modules, disabling the connection, etc. These endpoints are available to the Jetpack dashboard (and to Calypso via some older forwarding). These endpoints don't need to be available for WordPress.com sites, since it doesn't make sense there (at least for now).
* `wpcom/v2` should be things dealing with the settings for the features in Jetpack: liking a post, creating a contact form, generating a post by email address, etc. These endpoints are available for both Jetpack sites and WordPress.com sites, since they make sense in both contexts.

According to that opinion, these Publicize endpoints would fit best under `wpcom/v2`.

tyxla commented on modules/publicize/publicize.php:
```diff
@@ -1075,6 +1101,191 @@ function publicize_checkbox_default( $checked, $post_id, $service_name, $connect
+		include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
+
+		// The `sharing` submenu page must exist for service connect URLs to be correct.
+		add_submenu_page( 'options-general.php', '', '', 'manage_options', 'sharing', '__return_empty_string' );
```
It would be so great if `get_available_service_data()` didn't assume it's being called in the wp-admin... not a blocker tho.

tyxla reviewed (commented):
Other than the concerns with the endpoint namespace and the missing `display_name` that's mentioned in https://github.com/Automattic/wp-calypso/pull/28102#pullrequestreview-168743801, this looks good and works well to me 👍

jeherve reviewed (dismissed):
I ran some tests (following https://github.com/Automattic/wp-calypso/pull/28102) and it seems to work well, except for a PHP notice I noted. The Publicize experience in Gutenberg could still be improved, but the functionality seems to work.

mdawaffe committed:
REST API: Publicize: Return `WP_Error` on failure.

This gives us the opportunity to be more informative, and the `WP_Error`
return value is used by all the other `permission_callback`s.

mdawaffe committed:
Typo

mdawaffe commented on modules/publicize/publicize.php:
```diff
@@ -442,6 +448,26 @@ function is_connecting_connection( $connection ) {
 	 * @return void
 	 */
 	function test_publicize_conns() {
+		wp_send_json_success( $this->get_publicize_conns_test_results );
```
Should be resolved as of 4ad3854f1.

mdawaffe commented:
I'd like to make several changes:
- [ ] Move most of the endpoint code to something like _inc/lib/core-api/wpcom-endpoints/publiciize.php and sync this with WP.com
- [ ] Switch to `wpcom/v2` namespace
- [ ] Rename some of the routes. For example, `/publicize/connections` does not just return a list of connections. It runs the connection tests for each one, which is expensive. Calypso is even calling the request `requestTestPublicizeConnections()` https://github.com/Automattic/wp-calypso/pull/28102/
- [ ] Rethink how the Publicize data is output in REST API Post requests. In old v1 JSON API, we output Publicize URLs for read requests, and accept some simpler inputs for write requests. Also, in the current APIs, by default we Publicize to everything. In this patch, we only Publicize to connections we explicitly turn on.

tyxla committed:
Add social logos dependency for now

tyxla committed:
Comment out broken test for now

tyxla committed:
Bump

mdawaffe commented:
Closing in favor of #10605.
